src/Visualization/Python/InterpolateVolumeData.py

```diff
@@ -0,0 +1,372 @@
+    source_volume_data: str, optional
+        the name of the .vol file inside the source file where the source data
+        can be found
+    target_volume_data: str, optional
+        the name of the .vol file inside the target file where the target data
+        is written
+    obs_start: float, optional
+        disregards all observation with observation_value before obs_start
+    obs_end: float, optional
+        disregards all observations with observation_value after obs_start
+    obs_stride: float, optional
+        will only take every obs_stride observation
+    """
+
+    if target_file_path is None:
+        target_file_path = source_file_path
+
+    source_file = spectre_h5.H5File(source_file_path, True)
+    source_vol = source_file.get_vol(source_volume_data)
+    dim = source_vol.get_dimension()
+
+    # here it is possible that source_file is opened by 2 objects,
+    # one of which is writing, seems to work though
+    target_file = spectre_h5.H5File(target_file_path, True)
+    target_file.insert_vol(target_volume_data, source_vol.get_version())
+    target_vol = target_file.get_vol(target_volume_data)
+
+    # for wrong mesh_args,
+    # the error message is actually pretty good so no need to catch it
+    target_mesh = Mesh[dim](*mesh_args)
+
+    # apply observation filter
+    observations = [
+        obs for obs in source_vol.list_observation_ids()
+        if obs_start <= source_vol.get_observation_value(obs) <= obs_end
+    ]
+    observations = observations[::obs_stride]
+
+    for obs in observations:
+        extents = source_vol.get_extents(obs)
+        bases = source_vol.get_bases(obs)
+        quadratures = source_vol.get_quadratures(obs)
+        tensor_names = source_vol.list_tensor_components(obs)
+        grid_names = source_vol.get_grid_names(obs)
+
+        if components_to_interpolate is not None:
+            tensor_names = list(
+                set(tensor_name for pattern in components_to_interpolate
+                    for tensor_name in tensor_names
+                    if re.match(pattern, tensor_name)))
+
+        # pre-load all tensors to avoid loading the full tensor for each element
+        tensors = [
+            np.array(source_vol.get_tensor_component(obs, name), copy=False)
+            for name in tensor_names
+        ]
+
+        volume_data = []
+        index = 0  # keeps track of position in tensor components
+        # iterate over elements
+        for i, extent in enumerate(extents):
+            source_mesh = Mesh[dim](extent, basis_from_string(bases[i]),
+                                    quadrature_from_string(quadratures[i]))
+
+            interpolant = RegularGrid[dim](source_mesh, target_mesh)
+            grid_points = source_mesh.number_of_grid_points()
+
+            tensor_comps = []
+            # iterate over tensors
+            for j, tensor in enumerate(tensors):
+                dataVector = DataVector(tensor[index:index + grid_points],
```

Comment:
You don't need to keep track of the `index`. Use `spectre_h5.offset_and_length_for_grid(grid_name, all_grid_names, all_extents)` instead.